pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -686,12 +839,46 @@ public SubscriptionStats getStats() {
 
     @Override
     public synchronized void redeliverUnacknowledgedMessages(Consumer consumer) {
-        dispatcher.redeliverUnacknowledgedMessages(consumer);
+        ConcurrentLongLongPairHashMap positionMap = consumer.getPendingAcks();
+        // Check if message is in pending_ack status.
+        if (null != positionMap) {
+            List<PositionImpl> pendingPositions = new ArrayList<>();
+            PositionImpl cumulativeAckPosition = null == this.pendingCumulativeAckMessage ? null :
+                    (PositionImpl) this.pendingCumulativeAckMessage;
+
+            positionMap.asMap().entrySet().forEach(entry -> {
+                PositionImpl position = new PositionImpl(entry.getKey().first, entry.getKey().second);
+                if ((pendingAckMessages == null || (pendingAckMessages != null &&
+                        !this.pendingAckMessages.contains(position))) &&
+                        (null == cumulativeAckPosition ||
+                                (null != cumulativeAckPosition && position.compareTo(cumulativeAckPosition) > 0))) {
+                    pendingPositions.add(position);
+                }
+            });
+
+            dispatcher.redeliverUnacknowledgedMessages(consumer, pendingPositions);
+        } else {
+            dispatcher.redeliverUnacknowledgedMessages(consumer);
+        }
     }
 
     @Override
     public synchronized void redeliverUnacknowledgedMessages(Consumer consumer, List<PositionImpl> positions) {
-        dispatcher.redeliverUnacknowledgedMessages(consumer, positions);
+        // Check if message is in pending_ack status.
+        List<PositionImpl> pendingPositions = new ArrayList<>();
```

Comment:
@MarvinCai 

I missed this in the previous reviews. We should avoid creating a new array list when this is called when transaction is not used.

So the logic should be something like:

```
List<PositionImpl> pendingPositions = positions;

if (pendingAckMessages != null || cumulativeAckPosition != null) {
    pendingPositions = new ArrayList<>();
   // process the logic here
}
```